pkg/storage/engine.go

```diff
@@ -810,17 +810,36 @@ type Writer interface {
 	// Writer implementations, this is a no-op.
 	LogLogicalOp(op MVCCLogicalOpType, details MVCCLogicalOpDetails)
 
-	// SingleClearEngineKey removes the most recent write to the item from the db
-	// with the given key. Whether older writes of the item will come back
-	// to life if not also removed with SingleClear is undefined. See the
-	// following:
-	//   https://github.com/facebook/rocksdb/wiki/Single-Delete
-	// for details on the SingleDelete operation that this method invokes. Note
-	// that clear actually removes entries from the storage engine, rather than
-	// inserting MVCC tombstones. This is a low-level interface that must not be
-	// called from outside the storage package. It is part of the interface
-	// because there are structs that wrap Writer and implement the Writer
-	// interface, that are not part of the storage package.
+	// SingleClearEngineKey removes the most recent write to the item from the
+	// db with the given key, using Pebble's SINGLEDEL operation. This
+	// originally resembled the semantics of RocksDB
+	// (https://github.com/facebook/rocksdb/wiki/Single-Delete), but was
+	// strengthened in Pebble such that sequences (from more recent to older)
+	// like SINGLEDEL#20, SET#17, DEL#15, ... work as intended since there has
+	// been only one SET more recent than the last DEL. These also work if the
+	// DEL is replaced by a RANGEDEL, since RANGEDELs are used extensively to
+	// drop all the data for a replica, which may then be recreated in the
+	// future. The behavior is non-deterministic and definitely not what the
+	// caller wants if there are multiple SETs/MERGEs etc. immediately older
+	// than the SINGLEDEL.
+	//
+	// Note that using SINGLEDEL requires the caller to not duplicate SETs
+	// without knowing about it. That is, the caller cannot rely simply on
+	// idempotent writes for correctness, if they are going to be later deleted
+	// using SINGLEDEL. A current case where duplication without knowledge can
+	// happen is sstable ingestion for "global" keys, say during import and
```

Comment:
It can happen on any ingestion. For example, `applySnapshot` ingests, and this ingestion has a good chance of being flushable. So it can be replayed more than once.

Snapshots include all the "replicated" keys, which include rangeID-local keys.

Reply:
Also, "replicated" keys include lock table keys: https://github.com/cockroachdb/cockroach/blob/8a962614c81684eb04e3f494aeb222896662b032/pkg/kv/kvserver/rditer/select.go#L97-L98.

So they end up in snapshots. Are these the keys for which we use `SingleDelete`?

Reply:
Taking it back. Doesn't happen on any ingestion. Only on ones that we have a chance to replay from raft logs after a restart.